src/librustc_mir/interpret/traits.rs

```diff
@@ -46,48 +46,48 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> InterpretCx<'a, 'mir, 'tcx, M>
         let ptr_align = self.tcx.data_layout.pointer_align.abi;
         // /////////////////////////////////////////////////////////////////////////////////////////
         // If you touch this code, be sure to also make the corresponding changes to
-        // `get_vtable` in rust_codegen_llvm/meth.rs
+        // `get_vtable` in `rust_codegen_llvm/meth.rs`.
         // /////////////////////////////////////////////////////////////////////////////////////////
         let vtable = self.memory.allocate(
-            ptr_size * (3 + methods.len() as u64),
+            ptr_size * (methods.iter().map(|l| 3 + l.len() as u64).sum()),
             ptr_align,
             MemoryKind::Vtable,
         );
         let tcx = &*self.tcx;
 
+        let mut cur_ptr = vtable;
```

Comment:
Everything below here in this function could be in a separate commit, as it's just a (very good!) cleanup.

Reply:
> Do you have some links to the related discussion, issues, RFCs?
> 
> Am I right in assuming that the current unfinished implementation will allow casting `dyn Foo` to `Bar` and `Boo` if `Foo: Bar + Foo`, even though one of these is definitely not possible because the vtables can't match? So the implementation strategy is solely for zero cost upcasting where the start of `Foo`'s vtable is equal to the start of `Bar`'s vtable?

I presume you meant `Boo: Bar + Foo`? This PR isn't allowing multi-trait objects though, so we don't need to worry about such cases unless one of `Foo` or `Bar` is an auto-trait, in which case the vtable isn't affected.

Anyway, for background, @nikomatsakis and I have discussed this feature informally for some time now, and we had a [conversation on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/144729-wg-traits/topic/object.20upcasting) a few weeks ago, which @eddyb also participated in – I believe Eddy tagged you in a question right at the end and you replied, so you might have seen some of it? He explains the vtable layout in any case (which I may have gotten wrong in this PR, but hopefully not).